api/app/outcome/api.py

```diff
@@ -0,0 +1,152 @@
+import traceback
+from flask_restful import reqparse, fields, marshal_with
+import flask_restful as restful
+from flask import g, request
+from sqlalchemy.exc import SQLAlchemyError
+
+from app.outcome.models import Outcome, Status
+from app.outcome.repository import OutcomeRepository as outcome_repository
+from app.events.repository import EventRepository as event_repository
+from app.users.repository import UserRepository as user_repository
+from app.utils.emailer import send_mail
+from app.email_template.repository import EmailRepository as email_repository
+
+from app.utils.auth import auth_required, event_admin_required
+from app import LOGGER
+from app import db
+from app.utils import errors
+from app.utils import misc
+
+
+def _extract_status(outcome):
+    if not isinstance(outcome, Outcome):
+        return None
+    return outcome.status.name
+
+outcome_fields = {
+    'id': fields.Integer,
+    'status': fields.String(attribute=_extract_status),
+    'timestamp': fields.DateTime(dt_format='iso8601'),
+}
+
+user_fields = {
+    'id': fields.Integer,
+    'email': fields.String,
+    'firstname': fields.String,
+    'lastname': fields.String,
+    'user_title': fields.String
+}
+
+outcome_list_fields = {
+    'id': fields.Integer,
+    'status': fields.String(attribute=_extract_status),
+    'timestamp': fields.DateTime(dt_format='iso8601'),
+    'user': fields.Nested(user_fields),
+    'updated_by_user': fields.Nested(user_fields)
+}
+
+
+class OutcomeAPI(restful.Resource):
+    @auth_required
+    @marshal_with(outcome_fields)
+    def get(self):
+        req_parser = reqparse.RequestParser()
+        req_parser.add_argument('event_id', type=int, required=True)
+        args = req_parser.parse_args()
+
+        user_id = g.current_user['id']
+        event_id = args['event_id']
+
+        try:
+            outcome = outcome_repository.get_latest_by_user_for_event(user_id, event_id)
+            if not outcome:
+                return errors.OUTCOME_NOT_FOUND
+            
+            return outcome
+
+        except SQLAlchemyError as e:
+            LOGGER.error("Database error encountered: {}".format(e))            
+            return errors.DB_NOT_AVAILABLE
+        except: 
+            LOGGER.error("Encountered unknown error: {}".format(traceback.format_exc()))
+            return errors.DB_NOT_AVAILABLE
+
+    @event_admin_required
+    @marshal_with(outcome_fields)
+    def post(self, event_id):
+        req_parser = reqparse.RequestParser()
+        req_parser.add_argument('user_id', type=int, required=True)
+        req_parser.add_argument('outcome', type=str, required=True)
+        args = req_parser.parse_args()
+
+        event = event_repository.get_by_id(event_id)
+        if not event:
+            return errors.EVENT_NOT_FOUND
```

Comment:
I think just another example here about how the global error handling can work to clean up the code. If we raise an `EventNotFoundError` in the `get_by_id()` method if the event doesn't exist, then we won't have to check if the event exists since the exception would have been raised if it's not there. The exception handler for my hypothetical `ApplicationError` would then take it, do some custom logic (like logging for example but won't be needed here) and then look at the class to find the appropriate message and status code and return that